### PR TITLE
Fix `IncompatibleClassChangeError` in Android instrumentation test (#1035)

### LINKED ISSUE #1035: java.lang.IncompatibleClassChangeError: Superclass my.project.TestInterface of my.project.TestInterface_1_Proxy is an interface (declaration of 'my.project.TestInterface_1_Proxy' appears in /data/user/0/my.project.test/app_dxmaker_cache/Generated_712335553.jar)
### Prerequisites

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Expected Behavior

I expect my android instrumentation tests to be running as they were on 1.13.3 (now failing on 1.13.4 with the exception described below)

### Current Behavior

When running an Android instrumented test that uses mockk and initializes dependencies with koin I get test failure with a stacktrace mentioned below

This happens with a very simple setup of trying to inject a class into a viewModel that I want to mock using mockk.
When I'm not trying to inject a class into the viewModel the test passes green.

#### Steps to Reproduce

1. With the setup mentioned below, run the androidTest and it will fail with the below mentioned stacktrace
```
java.lang.IncompatibleClassChangeError: Superclass my.project.TestInterface of my.project.TestInterface_1_Proxy is an interface (declaration of 'my.project.TestInterface_1_Proxy' appears in /data/user/0/de.my.project.test/app_dxmaker_cache/Generated_712335553.jar)
```

#### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* MockK version: 1.13.4 (wasn't happening on 1.13.3)
* OS: Android API33
* Kotlin version: 1.8.0
* JDK version: 11.0.13
* JUnit version: 4.13.2
* Type of test: android instrumented test

#### Stack trace

```
java.lang.IncompatibleClassChangeError: Superclass my.project.TestInterface of my.project.TestInterface_1_Proxy is an interface (declaration of 'my.project.TestInterface_1_Proxy' appears in /data/user/0/de.my.project.test/app_dxmaker_cache/Generated_712335553.jar)
at java.lang.VMClassLoader.findLoadedClass(Native Method)
at java.lang.ClassLoader.findLoadedClass(ClassLoader.java:738)
at java.lang.ClassLoader.loadClass(ClassLoader.java:363)
at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
at com.android.dx.stock.ProxyBuilder.loadClass(ProxyBuilder.java:358)
at com.android.dx.stock.ProxyBuilder.buildProxyClass(ProxyBuilder.java:340)
at io.mockk.proxy.android.transformation.AndroidSubclassInstrumentation.subclass(AndroidSubclassInstrumentation.kt:37)
at io.mockk.proxy.common.ProxyMaker.subclass(ProxyMaker.kt:148)
at io.mockk.proxy.common.ProxyMaker.proxy(ProxyMaker.kt:54)
at io.mockk.impl.instantiation.JvmMockFactory.newProxy(JvmMockFactory.kt:34)
at io.mockk.impl.instantiation.AbstractMockFactory.newProxy$default(AbstractMockFactory.kt:24)
at io.mockk.impl.instantiation.AbstractMockFactory.mockk(AbstractMockFactory.kt:59)
at my.project.TestTest$koinTestRule$1$1$1.invoke(TestTest.kt:58)
at my.project.TestTest$koinTestRule$1$1$1.invoke(TestTest.kt:28)
... [the stacktrace continues but doesn't seem to relevant after this point]
```
#### Minimal reproducible code (the gist of this issue)

The relevant classes (simplified) below outline the setup in which this is reproducible

koin versions:
```
val koinCoreVersion = "3.3.2"
api("io.insert-koin:koin-core:$koinCoreVersion")
api("io.insert-koin:koin-test:$koinCoreVersion")
api("io.insert-koin:koin-test-junit4:$koinCoreVersion")
api("io.insert-koin:koin-test-junit5:$koinCoreVersion")

val koinAndroidVersion = "3.3.2"
api("io.insert-koin:koin-android:$koinAndroidVersion")
api("io.insert-koin:koin-androidx-navigation:$koinAndroidVersion")

api("io.insert-koin:koin-androidx-compose:3.4.1")
```

Test class:
```kotlin
class TestTest : KoinTest {
    @get:Rule(order = 1)
    val koinTestRule = KoinTestRule.create {
        modules(
            module {
                single<TestInterface> { mockk(relaxed = true) }
            },
            testModule
        )
    }

    @get:Rule(order = 2)
    val composeTestRule = createAndroidIntentComposeRule<TestActivity> {
        Intent(it, TestActivity::class.java)
    }

    @Test
    fun test_test() {
        composeTestRule.onNodeWithTag("test_tag").assertExists()
    }
}
```

The Activity:
```kotlin
class TestActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            val viewModel: TestViewModel = getViewModel()
            Text(text = "text", modifier = Modifier.testTag("test_tag"))
        }
    }
}
```

ViewModel:
```kotlin
class TestViewModel(testInterface: TestInterface) : ViewModel()
```

The interface I want to inject:
```kotlin
interface TestInterface
```

The koin module:
```kotlin
val testModule = module {
    viewModel {
        TestViewModel(get())
    }
}
```

### PR DESCRIPTION
**Pull request by LeonRa**
## Summary

Fixes #1035

When eliminating the duplicate `ProxyMaker` in https://github.com/mockk/mockk/pull/1025 (cc @aSemy) it appears that some minor differences between the vanilla and Android JVMs were lost, resulting in the failure of Android instrumentation tests. These were not caught early on because it appears that mockk Android instrumentation tests were not running as part of CI.

This PR does the following:

- Picks up the instrumentation test fix proposed by @kubode in https://github.com/mockk/mockk/pull/1129
- Reverts the breaking https://github.com/mockk/mockk/pull/1025 as suggested by @oliverspryn in [this comment](https://github.com/mockk/mockk/issues/1035#issuecomment-1688479891)

All kudos goes to the above individuals for suggesting fixes, I merely combined them!

## Verification

- [X] Locally running mockk's [Android instrumentation test for abstract class mocks](https://github.com/mockk/mockk/blob/2e186ea81d2c0b749a307d42c77e0795dd0542d3/modules/mockk-agent-android/src/androidTest/java/io/mockk/proxy/android/AndroidMockKProxyMakerTest.java#L118-L130) no longer fails
- [X] A project consuming MockK no longer experiences the `IncompatibleClassChangeError` in Android instrumentation tests

### CODE REVIEW

**Comment by aSemy** [comment]
Oops, sorry for this! Thanks for hunting down and fixing it though.

It wasn't clear that the differences between the two classes were intentional - I assumed that one had been updated and the other had been forgotten. If you think others might make the same mistake as me, can you add some comments to the classes to clarify, and hopefully prevent a similar mistake in the future?

**Comment by Raibaz** [comment]
Thanks a lot for looking into this!

`JvmMockKProxyMakerTest[jvm] > garbageCollectedProxy()[jvm] FAILED`

Can you please fix this failing test?

**Comment by LeonRa** [comment]
@aSemy No worries at all, this has has a lot of moving parts. As I mentioned in the description, I also think of this as a systems problem - the tests were not running, so you couldn't know that something went wrong. I don't think a repeat is particularly likely, so I won't add comments here, but I also think what you were doing is the right direction here. I'd be interested in trying to merge these again to understand / document what the fine differences are

**Comment by LeonRa** [comment]
@Raibaz I noticed that the previous results also had a failure on Android API 21, which I'm not sure it makes sense to fix as part of this work. An argument could likely be made for upping the minimum supported version as API 21 is no longer that mainstream as well, but that's likely a discussion for a separate PR

**Comment by Raibaz** [comment]
I agree with you, I think we can drop support for API level 21 in tests.

Would you be able to send out a PR removing support for API 21 (and maybe adding some newer ones), so that we can get that merged, then rebase this branch on top of the new master and merge this?

Thanks!

**Comment by LeonRa** [comment]
@Raibaz [Here you go!](https://github.com/mockk/mockk/pull/1146) Also rebased this PR on top of that branch as requested, so if all is well in CI here we should be okay to merge sequentially 🤞

**Comment by LeonRa** [comment]
Okay, hopefully last time is the charm. I've rolled back the addition of API 31 as it looks like there isn't a valid x86 target for it. Opting to defer adding new targets to a future PR so that we can get this fix shipped

**Review by osipxd — COMMENTED** [review]
I'm not a real reviewer, just passing by and noticed a few stylistic remarks 🙃

**Comment by Raibaz** [comment]
Thanks @osipxd for spotting these!

They all make sense to me.

@LeonRa can you please fix the remarks @osipxd made? After that I think we can merge this.

Thank you both!

File: modules/mockk-agent/src/jvmMain/kotlin/io/mockk/proxy/jvm/ProxyMaker.kt [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        return if (inliner != null) {
+            val transformRequest = TransformationRequest(superclasses, SIMPLE)
+
+            inliner.execute(transformRequest)
+        } else {
+            if (!Modifier.isFinal(clazz.modifiers)) {
+                warnOnFinalMethods(clazz)
+            }
+
+            {}
```
[the remark this answers is not in the document]
Comment (LeonRa) [reply modules/mockk-agent/src/jvmMain/kotlin/io/mockk/proxy/jvm/ProxyMaker.kt:96]: I believe these are returning an empty lambda as the function return is `() -> Unit`, so they're not redundant, but I can add a comment to make that more clear

**Comment by onXAdamBrown** [comment]
Any idea when this might be in a release?

**Comment by LeonRa** [comment]
> Any idea when this might be in a release?

Wondering this as well. I was hoping we'd cut a build prior to all of the dependency bumps that have come in since to reduces the chances of other breakages

**Comment by Raibaz** [comment]
Planning to publish a release later this week, likely tomorrow or Friday

**Pull request merged**